### PR TITLE
feat: Update Polygon from `MATIC` to `POL`

### PR DESCRIPTION
**Pull request by gambinish**
## **Description**

Add `POL` token image, and maps it to Polygon Network config. See [blog post](https://polygon.technology/blog/save-the-date-matic-pol-migration-coming-september-4th-everything-you-need-to-know) for more context.

Adds chain collision check to not flag `POL` as scam token

Adds migration to overwrite `MATIC` ticker to `POL`

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/26671?quickstart=1)

## **Related issues**

Fixes: https://consensyssoftware.atlassian.net/browse/MMASSETS-362

## **Manual testing steps**

1. When adding Polygon as a network from `FEATURED_RPCS`, ticker symbol should be POL
2. When switching to Polygon Mainnet, ticker should be POL (not MATIC)
2. Can import `Polygon Network Token` on Ethereum Mainnet
3. Can import `Matic Network Token` on Ethereum Mainnet (backward compatibility)

> Note, that both `MATIC` and `POL` erc20 tokens can already be imported on Ethereum Mainnet, so this behavior should remain unchanged.

Additionally, when a user upgrades to this build, migration should run and overwrite ticker from MATIC to POL:

1. Install a previous build (checkout `develop` and `yarn && yarn start` should be fine here) -> Polygon Network should show `MATIC` as ticker. You may need to remove and re-add Polygon Network, as migrations are incremental.
2. checkout to this branch, `yarn && yarn start`, migration 128 should run (Running migration 128... in console) -> Polygon Network should show `POL` as ticker

## **Screenshots/Recordings**

Before:

Screenshot 2024-08-27 at 2 39 15 PM
Screenshot 2024-08-27 at 2 39 36 PM

After:

Screenshot 2024-08-27 at 1 47 40 PM
Screenshot 2024-08-27 at 1 48 11 PM

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/migrations/128.ts
Comment (bergeron) [inline app/scripts/migrations/128.ts:29]: as you discovered in the mobile migration, we may need to update providerConfig.ticker iff polygon is the current network.  Because this will be a fix for 12.1.x where we still have the old network controller.  You can test the migration against that branch https://github.com/MetaMask/metamask-extension/tree/Version-v12.1.0

I think we'd still merge it to develop first, then cherry pick to a 12.1.x branch.  We may also tweak the migration number to slot into that release branch more conveniently.

Comment (gambinish) [reply app/scripts/migrations/128.ts:30]: Updated migration to handle legacy `providerConfig` as well as a couple other cleanup things.

If we do cherry pick this from develop onto `12.1.1` I might need some hand holding to make sure I do this correctly. Not super familiar with this cherry-picking process.

File: app/scripts/migrations/128.ts
Comment (Gudahtt) [inline app/scripts/migrations/128.ts:1]: When we're writing a migration that we intend to backport to an older release, typically we number the migration so that it doesn't need to be re-numbered when we cherry-pick it. In this case we're targeting Version-v12.1.1 which is on migration 121, so we'd be unable to use 128 for this migration. If we did, migrations 122-127 would be skipped when they shipped in alter releases.

Instead we can call this migration 121.1, so that it runs in the correct order without us needing to re-number anything else

File: shared/constants/network.ts
Comment (Gudahtt) [inline shared/constants/network.ts:379]: I'm still seeing various references to this image in test files and Storybook, and the image itself remains present. Should we we remove the image and the remaining references? It would be nice to remove the extra image for the sake of reducing bundle size.

Comment (gambinish) [reply shared/constants/network.ts:379]: Sure I can remap these references.

File: app/scripts/migrations/128.ts
Comment (gambinish) [reply app/scripts/migrations/128.ts:1]: I did notice that when I checked out the previous build, and had a feeling this would be the case.

I'll go ahead and make that change.

File: shared/constants/network.ts
Comment (gambinish) [reply shared/constants/network.ts:379]: https://github.com/MetaMask/metamask-extension/pull/26671/commits/d75c312097bb719a6a0dad166d775636ab5c716a

File: app/scripts/migrations/128.ts
Comment (gambinish) [reply app/scripts/migrations/128.ts:1]: https://github.com/MetaMask/metamask-extension/pull/26671/commits/46688c0135ddb7164e91f5587750b45174f9652b

File: app/scripts/migrations/121.1.ts
Comment (Gudahtt) [inline app/scripts/migrations/121.1.ts:30]: Nit: Two suggestions:

* Ideally this function would either mutate the inputs, or return the transformed state without mutating the inputs, not do both. Currently the return value is not used, so maybe we're better off returning nothing here so that this is more clear.
* It's generally recommended to explicitly document return types rather than relying on TypeScript inference, as it's easier then for readers to understand what the function does, and it highlights changes to return types that might go unnoticed if we rely solely on type inference (see here for more details: https://github.com/MetaMask/contributor-docs/blob/main/docs/typescript.md#for-functions-and-methods-provide-explicit-return-types)

File: app/scripts/migrations/121.1.ts
Comment (Gudahtt) [inline app/scripts/migrations/121.1.ts:1]: I think we'll need to migrate the `CurrencyController` state as well. The `CurrencyController.currencyRates` state is keyed by ticker symbol.

This is from the CurrencyRateController in core (for some reason the state property omits "Rate" from the name).

Comment (Gudahtt) [reply app/scripts/migrations/121.1.ts:1]: This doesn't seem overly important 🤔 It would get refreshed right away if we forgot to migrate it, and this state is tiny and not doing much harm if we leave it around (and in terms of saving space, we really ought to clear out old rates in a more general way, not just obsolete ones).

Maybe nice to include in this migration? But I'm not sure how much it matters.

**Review by Gudahtt — COMMENTED** [review]
Looks great! I'll do a smoke test tomorrow and approve if I don't find anything else, if it's still open by then

File: app/scripts/migrations/121.1.ts
Comment (Gudahtt) [inline app/scripts/migrations/121.1.ts:65]: This line has no effect, and can be omitted

```suggestion
```

The state changes are applied already by this point. And even if they weren't, this would have no effect because we're creating a new object here (which then is not returned or accessed ever).

Comment (gambinish) [reply app/scripts/migrations/121.1.ts]: Thanks. This change is in this commit here: https://github.com/MetaMask/metamask-extension/pull/26671/commits/d9093413eb39dd1e24f322e8898cf205e962c50d

**Comment by garrettbear** [comment]
It looks good for DS on the changes, but I see there are a couple of comments from @Gudahtt

File: app/scripts/migrations/121.1.ts
Comment (Gudahtt) [inline app/scripts/migrations/121.1.ts:51]: Huh, why is there an `imageUrl` in the `rpcPrefs` state 🤔 This is missing from the `NetworkController` state type.

Comment (Gudahtt) [reply app/scripts/migrations/121.1.ts:52]: It's there, so we'll need to migrate it, but I think we should follow up to remove this state completely. It looks like we don't need it, and it's introducing a type error (that will cause problems as we make progress with the TypeScript migration).

It looks like it is only in state because it leaked here from a constant. But it's not set dynamically, so there is no need for it to be in state. We can refactor places where this state is used to reference a constant instead. Or we can add it to the NetworkController type, if we want to start setting it dynamically as well.

Comment (gambinish) [reply app/scripts/migrations/121.1.ts:52]: Sounds good. I may follow up with the Assets team here since I know there are a lot of NetworkController upgrades in flight.

Comment (bergeron) [reply app/scripts/migrations/121.1.ts:52]: I have looked into this as part of my network controller upgrade and think I know what's going on.  Images were never supported in the network controller state, but leaked in somehow.  In my PR I move everything to grab them from constants.  We never supported a way for users to provide custom network images.

Comment (gambinish) [reply app/scripts/migrations/121.1.ts:52]: Sounds like @bergeron has already dealt with this on the `NetworkController` upgrade work being done

File: app/scripts/migrations/121.1.ts
Comment (gambinish) [reply app/scripts/migrations/121.1.ts:1]: Are you cool if we punt the `CurrencyController` upgrade to another migration? If it's not overly important, I can capture this in another ticket to think about in more detail.

File: app/scripts/migrations/121.1.ts
Comment (bergeron) [reply app/scripts/migrations/121.1.ts:52]: Yes, but to be clear it likely still needs to be migrated as of whichever 12.1.x this goes into.  So I'd keep the code here.

**Comment by martahj** [comment]
A potential concern: I'm not able to initiate a swap _from_ MATIC. I could see preventing users from swapping _to_ MATIC, but it would be helpful if users could use MM Swaps to exit out of their MATIC position into POL (or another token of their choice); in fact, we were going to waive our fee for users to do this to drive usage. Wdyt about continuing to allow users to swap out of MATIC?

Screenshot 2024-08-29 at 12 23 27 PM

File: app/scripts/migrations/121.1.ts
Comment (Gudahtt) [reply app/scripts/migrations/121.1.ts:1]: Sure, sounds good

Comment (gambinish) [reply app/scripts/migrations/121.1.ts:1]: Tracking here: https://consensyssoftware.atlassian.net/browse/MMASSETS-371

File: app/scripts/migrations/121.2.test.ts
Comment (Gudahtt) [inline app/scripts/migrations/121.2.test.ts:26]: Nit: It's possible that this test might pass even when changes were made to state, because the assertion is using the same object reference as we're passing in. So if it were mutated, it still would match.

Our migration copy pasta includes a clone step, so in practice this is guaranteed not to happen, but if we mistakenly remove that clone step the test might pass when it should fail. To guard against this, I'd recommend using `cloneDeep` to clone the state inputs in each of these tests.

Comment (gambinish) [reply app/scripts/migrations/121.2.test.ts]: https://github.com/MetaMask/metamask-extension/pull/26671/commits/e8491d8b23f9f2ed08f48a160b197c161a411cc1

**Review by Gudahtt — APPROVED** [review]
LGTM!

File: ui/components/app/add-network/add-network.stories.js
Comment (garrettbear) [inline ui/components/app/add-network/add-network.stories.js:4]: i wonder if this was even working before since it should have been `matic-token.png` and that was set to `svg` 🤔

Comment (Gudahtt) [reply ui/components/app/add-network/add-network.stories.js:4]: It looks like we had both formats previously. I see that the old PNG logo is deleted in this diff, and the SVG is renamed

Comment (gambinish) [reply ui/components/app/add-network/add-network.stories.js:4]: From what I can tell, these were all defined as svg on `develop`

https://github.com/MetaMask/metamask-extension/blob/develop/ui/components/app/add-network/add-network.stories.js

**Review by garrettbear — APPROVED** [review]
LGTM 
Screenshot 2024-08-29 at 1 33 34 PM

**Review by vinnyhoward — APPROVED** [review]

**Comment by vinnyhoward** [comment]
LGTM✅

**Review by bschorchit — APPROVED** [review]
Approving on behalf of confirmations to unblock this per [this message](https://consensys.slack.com/archives/C1L7H42BT/p1724965108791619?thread_ts=1724963150.129119&cid=C1L7H42BT) ("confirmations team is only a codeowner because of an affected test file, but the changes in that file were just related to test fixtures.")

**Pull request merged**
